src/java.base/share/classes/java/lang/Class.java

```diff
@@ -4348,7 +4348,7 @@ public String descriptorString() {
             return Wrapper.forPrimitiveType(this).basicTypeString();
 
         if (isArray()) {
-            return "[" + componentType.descriptorString();
+            return "[".concat(componentType.descriptorString());
```

Comment:
This could be optimized further for multi-dimensional arrays:
```
if (isArray()) {
  int arrayDepth = 1;
  Class<?> ct = componentType;
  while (ct.isArray()) {
    arrayDepth++;
    ct = ct.componentType;
  }
  return "[".repeat(arrayDepth).concat(ct.descriptorString());
```

Reply:
But isn't `componentType.descriptorString()` does this itself? Also multi-dimensional arrays are quite an infrequent usecase, aren't they?

Reply:
Yeah, it's just an optimization. Current code wouldbuild "[[..[[LFoo;" by recursively going down to `Foo`, build and return "LFoo;", then do "[" + "LFoo;", and so on. Infrequent enough that we can ignore it, sure, but since it's effectively reducing the algorithmic complexity from O(n*m) to O(n+m) I should at least mention it.

Reply:
I see that `TestPrimitiveArrayCriticalWithBadParam` constantly fails, so I'd probably revert this change